### ISSUE TITLE
"bad time"error from Telegram X client

### ISSUE THREAD

**Issue by terrytw** [9seconds/mtg#155]
I have been trying to set up a mtproxy server using mtg, and encountered this problem:

After mtg started running, I tried to connect to it using a Telegram X android client, it was successful, however after a few minutes, the client was unable to connect. At the same time, server log gave this warning:

`{"level":"warn","ts":1595240548.4428384,"logger":"proxy","msg":"Cannot perform client handshake","connection_id":"dc174eaae139afab","error":"failed tls handshake: bad time"}`

Client gave an `error 400 response hash mismatch`

I saw https://github.com/9seconds/mtg/issues/132, so I have been experimenting a little bit. First of all, since you mentioned it was a bug introduced in a commit in Feb, I tried a Telegram X version from Jan, and it produced the same error. 

Then I tried to use the official mtproxy code to build a binary, that works flawlessly. 

Then I tried mtg with the latest Telegram android client (non-X version), and it works flawlessly too. 

I also tried the python version (https://github.com/alexbers/mtprotoproxy)
And it works just fine. 

In summary:
Telegram + mtg =fine
Telegram X + original mtproxy = fine
Telegram X + python mtproxy = fine
Telegram X (old or new verison) + mtg = bad time error

P.S. This problem only occurs when using faketls mode. 
Also, successfully connecting to telegram server via other functioning proxy or VPN would somehow reset the situation, making this proxy usable again (for a couple of minutes, until same situation happens again).

**Comment by terrytw** [comment]
I just saw https://github.com/alexbers/mtprotoproxy/issues/223

So I tried to disable anti-replay function by adding `--anti-replay-max-size 0`
And the problem persists.

**Comment by 9seconds** [comment]
I do not have TelegramX installed but I can try to check what is happening later

**Comment by terrytw** [comment]
Much appreciated! 

Just a side note, I tried this guy's (https://github.com/9seconds/mtg/issues/147) mtg server, it produced the same error, so it's not anything related with my server setup.

**Comment by 9seconds** [comment]
I've checked with faketls and latest version of Telegram X (0.22.8.13610). Works fine. Cannot reproduce that issue. Could you please try to update and check again?

**Comment by terrytw** [comment]
Thank you for checking on it!
I have tried 1.0.6 version, same error pops up. 
I am using Telegram X 0.22.8.1361. I have tried 3 different devices, with Android 8, 9, and 10 respectively.  (Carbon ROM, Pixel Experience, and Oxygen OS)
I have tried different servers as well. 
I am quite surprised that you cannot reproduce it....It only took me less than 5 minutes before the errors start to pop up.

**Comment by 9seconds** [comment]
Can it be an issue with a time on your mtg host? Could you please check logs if there any messages about time drift etc?

**Comment by terrytw** [comment]
Here is the complete logs of a newly established host, when `"error":"failed tls handshake: bad time"` starts to flush the screen, problem emerges. 
![1](https://user-images.githubusercontent.com/24227628/89902729-654c1700-dc19-11ea-8730-53323ccb8e1c.PNG)


I am not sure if it means anything: 
On client side, it shows `first part of response to hello is invalid` now (1.0.5 and 1.0.6), it used to be `error 400 response hash mismatch`

**Comment by terrytw** [comment]
Can you please try using Telegram X for longer period of time with mtg? For example 30 minutes. 

If you still cannot reproduce this problem, then feel free to close this issue. 

Thank you for your time!

**Comment by 9seconds** [comment]
I'm trying it now. Can you please use a build from this PR/branch: https://github.com/9seconds/mtg/pull/157? If necessary, I can build binaries for you. This build adds verbosity to that 'bad time' message.

**Comment by 9seconds** [comment]
The thing is that after decoding TLS payload we've verified that magic bytes are fine so this should purely be a time skew problem. This error can be triggered in 2 cases:

1. If your OS does not give a correct unix time to application so only "time from boot" is available
2. If time difference is more than 5 seconds.

1st is really rare nowadays. I expect that some clock is skewed. In order to do that we need to find where time is incorrect. Could you please repeat an experiment with a branch above and check a time when it happened? It could be that server clocks are skewed, it could be that client clocks are skewed.

**Comment by terrytw** [comment]
Yes, I am building right now, will report to you.

**Comment by terrytw** [comment]
Just ran test with new build, this is the result:
![1](https://user-images.githubusercontent.com/24227628/90003468-b5cd7e00-dcc6-11ea-85f7-e936bf18fa7a.PNG)

**Comment by terrytw** [comment]
I think it's client's time that is skewed. Just did another test:
![1](https://user-images.githubusercontent.com/24227628/90004337-07c2d380-dcc8-11ea-9fa3-0a94bf749bc5.PNG)
In the end I got the server's time stamp with `date +%s`, and confirmed with a timestamp webpage:
![2](https://user-images.githubusercontent.com/24227628/90004432-39d43580-dcc8-11ea-8ef2-4d8aac77398f.PNG)

I'm guessing that extracted timestamp comes from client? 
Server time was 1597227226 in the log, 1597227228 from `date +%s`, webpage shows 1597227229. 
Client timestamp was 1597227104

However since I tried it on 3 devices, with different Android versions and different rom as mentioned above (Android 8, 9, and 10 respectively. Carbon ROM, Pixel Experience, and Oxygen OS), it is unlikely it's a system issue. It still is related to Telegram X. 
But weird part is that Telegram X works fine with original mtproxy with faketls, and works fine with python version as well. (Just tested today)

**Comment by 9seconds** [comment]
IIRC, original proxies allow bigger skewness. I'm also do not remember if they verify timestamps or not at all.

To be honest, I no longer know how to help here :/

**Comment by terrytw** [comment]
At least we tried...Thanks!

**Comment by terrytw** [comment]
Update:

I reported to Telegram X's developer, and they said it was a known issue, but they did not know it also affect faketls mtproxy. 

Sorry for wasting your time, I thought it was mtg's issue because original and python works for me...

**Comment by 9seconds** [comment]
Oh, good to know. Thank you a lot!
